src/brisk.jl

```diff
@@ -0,0 +1,109 @@
+```
+
+| Argument | Type | Description |
+|----------|------|-------------|
+| `pattern_scale` | `Float64` | Scaling factor for the sampling window | 
+"""
+type BRISK{S, T, O} <: Params
+    threshold::Float64
+    octaves::Int
+    pattern_scale::Float64
+    pattern_table::Vector{Vector{S}}
+    smoothing_table::Vector{T}
+    orientation_weights::Vector{O}
+    short_pairs::Vector{O}
+    long_pairs::Vector{O}
+end
+
+typealias OrientationPair Vector{Int}
+typealias SamplePair Vector{Float64}
+
+function BRISK(; threshold::Float64 = 0.25, octaves::Int = 4, pattern_scale = 1.0)
+    pattern_table, smoothing_table = _brisk_tables(pattern_scale)
+    orientation_weights = OrientationPair[]
+    short_pairs = OrientationPair[]
+    long_pairs = OrientationPair[]
+    dminsq = (brisk_dmin * pattern_scale) ^ 2
+    dmaxsq = (brisk_dmax * pattern_scale) ^ 2
+    for i in 2:brisk_points
+        for j in 1:i - 1
+            dy, dx = pattern_table[1][j] - pattern_table[1][i]
+            norm = dy ^ 2 + dx ^ 2
+            if norm > dminsq
+                push!(long_pairs, OrientationPair([j, i]))
+                push!(orientation_weights, OrientationPair([round(Int, dy * 4096 / norm), round(Int, dx * 4096 / norm)]))
+            elseif norm < dmaxsq
+                push!(short_pairs, OrientationPair([j, i]))
+            end
+        end
+    end
+    BRISK(threshold, octaves, pattern_scale, pattern_table, smoothing_table, orientation_weights, short_pairs, long_pairs)
+end
+
+function _brisk_orientation{T<:Gray}(int_img::AbstractArray{T, 2}, keypoint::Keypoint, pattern::Array{SamplePair}, 
+                                        orientation_weights::Array{OrientationPair}, sigmas::Array{Float64}, long_pairs::Array{OrientationPair})
+    direction_sum_y = 0.0
+    direction_sum_x = 0.0
+    for (i, o) in enumerate(long_pairs)
+        offset_1 = pattern[o[1]]
+        offset_2 = pattern[o[2]]
+        intensity_diff = ImageFeatures._freak_mean_intensity(int_img, keypoint, offset_1, sigmas[o[1]]) - ImageFeatures._freak_mean_intensity(int_img, keypoint, offset_2, sigmas[o[2]])
+        direction_sum_y += orientation_weights[i][1] * intensity_diff / 4096
+        direction_sum_x += orientation_weights[i][2] * intensity_diff / 4096
+    end
+    angle = atan2(direction_sum_y, direction_sum_x)
+    scaled_angle = ceil(Int, (angle + pi) * brisk_orientation_steps / (2 * pi))
+    scaled_angle
+end
+
+function _brisk_tables(pattern_scale::Float64)
+    pattern_table = Vector{SamplePair}[]
+    smoothing_table = Vector{Float64}[]
+    for ori in 0:brisk_orientation_steps - 1
+        theta = ori * 2 * pi / brisk_orientation_steps 
+        pattern = SamplePair[]
+        sigmas = Float64[]
+        for (i, n) in enumerate(brisk_num_circular_pattern)
+            for circle_number in 0:n - 1
+                angle = (circle_number * 2 * pi / n) + theta
+
+                push!(pattern, SamplePair([brisk_radii[i] * sin(angle) * pattern_scale * 0.85, 
```

Comment:
Made the modifications just to check the benchmark scores - 

For the `BRISK()` call - 
```
# With Vector 

  samples:          238
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  16.66 mb
  allocs estimate:  350205
  minimum time:     14.51 ms (0.00% GC)
  median time:      22.13 ms (30.47% GC)
  mean time:        21.08 ms (26.61% GC)
  maximum time:     28.05 ms (32.27% GC)

# With Tuple

  samples:          694
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  1.15 mb
  allocs estimate:  12339
  minimum time:     6.54 ms (0.00% GC)
  median time:      7.02 ms (0.00% GC)
  mean time:        7.20 ms (2.53% GC)
  maximum time:     12.12 ms (28.43% GC)
```
For the `create_descriptor` call - 
```
# With Vector

  samples:          5
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  299.36 mb
  allocs estimate:  18317758
  minimum time:     998.75 ms (3.71% GC)
  median time:      1.01 s (3.64% GC)
  mean time:        1.02 s (3.58% GC)
  maximum time:     1.04 s (3.75% GC)

# With Tuple

  samples:          6
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
  memory estimate:  274.89 mb
  allocs estimate:  17556044
  minimum time:     996.81 ms (2.77% GC)
  median time:      999.79 ms (2.87% GC)
  mean time:        999.19 ms (2.88% GC)
  maximum time:     1.00 s (3.00% GC)
```

I guess there is quite an improvement in the memory storage in the `BRISK` call (I wasn't expecting much in `create_descriptor` anyways as it wasn't affected much by the conversion to `Tuple`). I am changing the `Vector` to `Tuple` :)